x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/CapacityResponseCacheTests.java

```diff
@@ -0,0 +1,256 @@
+
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.is;
+
+public class CapacityResponseCacheTests extends AutoscalingTestCase {
+    private ThreadPool threadPool;
+    // randomly run on generic pool, since it has at least 4 threads.
+    private final Consumer<Runnable> runOnThread = randomBoolean()
+        ? run -> threadPool.executor(ThreadPool.Names.MANAGEMENT).execute(run)
+        : run -> threadPool.executor(ThreadPool.Names.GENERIC).execute(run);
+
+    @Before
+    public void createThreadPool() {
+        threadPool = new TestThreadPool(getClass().getName());
+    }
+
+    @After
+    public void terminateThreadPool() {
+        terminate(threadPool);
+        threadPool = null;
+    }
+
+    public void testSingleResponse() throws ExecutionException, InterruptedException {
+        Queue<Long> responses = new ArrayDeque<>();
+        CapacityResponseCache<Long> cache = new CapacityResponseCache<>(runOnThread, cancelled -> responses.remove());
+        long response = randomLong();
+        responses.add(response);
+        PlainActionFuture<Long> future = new PlainActionFuture<>();
+
+        cache.get(() -> false, future);
+
+        assertThat(future.get(), equalTo(response));
+    }
+
+    public void testSingleCancel() {
+        CyclicBarrier barrier = new CyclicBarrier(2);
+        CapacityResponseCache<Long> cache = new CapacityResponseCache<>(runOnThread, cancelled -> {
+            await(barrier);
+            await(barrier);
+            cancelled.run();
+            return 0L;
+        });
+
+        AtomicBoolean cancelled = new AtomicBoolean();
+        PlainActionFuture<Long> future = new PlainActionFuture<>();
+        cache.get(cancelled::get, future);
+
+        await(barrier);
+        cancelled.set(true);
+        await(barrier);
+
+        expectThrows(TaskCancelledException.class, future::actionGet);
+    }
+
+    public void testMultipleRequests() throws Exception {
+        Queue<Supplier<Integer>> responses = new ArrayDeque<>();
+        CapacityResponseCache<Integer> cache = new CapacityResponseCache<>(runOnThread, ensureNotCancelled -> responses.remove().get());
+
+        CyclicBarrier barrier = new CyclicBarrier(2);
+        responses.add(() -> {
+            await(barrier);
+            await(barrier);
+            return 0;
+        });
+
+        PlainActionFuture<Integer> blockingFuture = new PlainActionFuture<>();
+        cache.get(() -> false, blockingFuture);
+        await(barrier);
+
+        int count = between(1, 10);
+        List<PlainActionFuture<Integer>> supersededFutures = new ArrayList<>(count + 1);
+        for (int i = 0; i < count; ++i) {
+            PlainActionFuture<Integer> supersededFuture = new PlainActionFuture<>();
+            supersededFutures.add(supersededFuture);
+            cache.get(() -> false, supersededFuture);
```

Comment:
Could we call this from various threads, not just sequentially?

Reply:
Absolutely, done in ed38bcb31f681d0297dec34f7b0bf9cd36a3c476, though your next comment revised some of the code changes in ed38bcb31f681d0297dec34f7b0bf9cd36a3c476